### DISCUSSION TITLE
Command "wsl --set-default-version 2" doesn't change version of WSL

### DISCUSSION THREAD

**Discussion by adanribeiro** [microsoft/WSL#6505] [General]
# Environment
Microsoft Windows [10.0.19042.746]
Windows 10 Pro, 20H2
Experience: Feature Experience Pack 120.2212.551.0

# Copy of the terminal output
On my Windows Terminal I type:
`wsl -l`
_Distribuições do Subsistema do Windows para Linux:
Ubuntu (Padrão)
Ubuntu-20.04_

Then, after run `wsl --set-default-version 2` I get:
_Para obter informações sobre as principais diferenças em relação ao WSL 2, visite https://aka.ms/wsl2_

What translated from pt-br mean "To get info about main differences relation to WSL 2, visit https://aka.ms/wsl2". But, unfortunelly, the default version keep on version_number 1. 

What possibly be wrong?

**Comment by therealkenc** [comment]
> But, unfortunelly, the default version keep on version_number 1.

Your default version is now 2. You can change the version of an already installed distro with [`--set-version`](https://docs.microsoft.com/en-us/windows/wsl/install-win10#set-your-distribution-version-to-wsl-1-or-wsl-2) not `--set-default-version`.

**Comment by adanribeiro** [comment]
OK, but when I run "wsl -l" it shows me that version "Ubuntu" is default (see the word "Padrão" on command output below).

```
Ubuntu (Padrão)
Ubuntu-20.04
```
Also, on this MS tutorial [here](https://docs.microsoft.com/pt-br/windows/wsl/install-win10) they said on Step 5 to use `wsl --set-default-version 2` to DEFINE the default version.

So it got a little complicated.

How can I confirm what version my system is running now or , at least, uninstall the version 1?

**Comment by therealkenc** [comment]
There is no version "the system" is running. Each of your installed distributions is either WSL1 or WSL2 independently, by-design. Both version 1 and 2 are always available; you don't "uninstall" WSL 1. You can get which version a given distribution is with `wsl.exe -l -v`. You can also find out your version from inside WSL with `cat /proc/version`. WSL1 will have a capitalized 'M' on 'Microsoft'.

On recent builds (possibly not yours) you can query the default version with `wsl.exe --status`. Looks like:

![image](https://user-images.githubusercontent.com/7049700/106504912-6b120f00-647c-11eb-930d-ab158a95e5b2.png)

Querying is a little bit superfluous. If you are ever unsure of your default, you can use `--set-default-version`, and then you know.

Perhaps the [documentation](https://github.com/MicrosoftDocs/WSL) could be improved to state explicitly that `--set-default-version` has no effect on existing distributions. Doc suggestions (or even better PRs) always welcome.

**Reply by adanribeiro** [reply]
When I run `wsl.exe -l -v` the output is:
```
  NAME            STATE           VERSION

* Ubuntu          Running         1
  Ubuntu-20.04    Stopped         2
```

What does it means?

**Reply by therealkenc** [reply]
It means your 'Ubuntu' distribution is on WSL version 1, and your 'Ubuntu-20.04' distribution is on WSL version 2. You can enter one or the other with `wsl.exe -d Ubuntu` and `wsl.exe -d Ubuntu-20.04`.

**Reply by JennEYoon** [reply]
Had this exact problem.  Thanks.

**Reply by MarvinXu** [reply]
> Each of your installed distributions is either WSL1 or WSL2 independently, by-design. Both version 1 and 2 are always available; you don't "uninstall" WSL 1. You can get which version a given distribution is with `wsl.exe -l -v`

Thank you for explaining! Very helpful! I'll add "You can change your distro's WSL version by `wsl --set-version <distribution name> <versionNumber>`"

**Comment by mateuszdrab** [comment]
I have an issue where wsl --set-default-version 2 simply ignores my input and returns the help text, the same happens when trying to convert a specific distro to version 2.

> C:\WINDOWS\system32>wsl --set-default-version 2
> Copyright (c) Microsoft Corporation. All rights reserved.
> 
> Usage: wsl.exe [Argument] [Options...] [CommandLine]
> 
> Arguments for running Linux binaries:
> ...

Build 21390, WSL reinstalled from scratch using wsl --install, latest kernel, there's nothing much that I can update.

Anyone experienced such issue?

**Reply by mccarthytyler** [reply]
I have experienced the same issue.  Have not yet found an answer.

**Reply by mccarthytyler** [reply]
I ended up installing a new distro (Ubuntu 20.04) which was already set to v2 (checked with `wsl -l -v`) and everything ended up working fine.  This is after running `wsl --set-default-version 2`

**Reply by mateuszdrab** [reply]
If `wsl --set-default-version 2` actually worked it would have been nice🤦‍♂️
Unfortunately, in my case it doesn't.

**Reply by sschuldenzucker** [reply]
Yeah, `--set-default-version` seems to do nothing, as does `--set-version`. Why is it always like this with MS?

**Reply by rodrigoaraujoti** [reply]
> If `wsl --set-default-version 2` actually worked it would have been nice🤦‍♂️ Unfortunately, in my case it doesn't.

Same problem here.

**Reply by crispd** [reply]
for me, `wsl --set-default-version 2` worked, but `wsl --set-version Debian` did not. (My distro's name is `Debian`)

**Reply by isCopyman** [reply]
same problem fuxx

**Reply by xaxnow** [reply]
`wsl.exe --set-version Ubuntu-22.04 2` works for me.

**Reply by isCopyman** [reply]
> same problem fuxx

Windows server 2016 and 2019 don't support wsl2 which causes invalid arguments

**Reply by pucoenraadmiddel** [reply]
I'm running Windows Server 2022 and have the same issue where wsl.exe --set-version Ubuntu-20.04 2 only returns the help text, not changing the version. I'm also not able to set the default to 2 before installing with the same behviour. 

Edition	Windows Server 2022 Standard
Version	21H2
Installed on	‎2023/‎02/‎19
OS build	20348.169

**Reply by mateuszdrab** [reply]
Is there an issue raised for this? If not, we need to raise one otherwise this will never get fixed/looked into.
I've experienced this myself long time ago, but it never got fixed until I eventually rebuilt the affected host.
I think the problem is actually distro instance specific.
@pucoenraadmiddel could you try installing another distro and try changing its version?

**Comment by pucoenraadmiddel** [comment]
Hi Mateusz,

Mine was an issue relating to a fresh install and the version of Windows Server 2022. When I updated it to the latest version it worked.

Check your version maybe?
